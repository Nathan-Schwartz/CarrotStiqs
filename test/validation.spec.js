const createCarrotStiqsClient = require('../src/carrotstiqs');
const { CarrotStiqsError } = require('../src/util');

const topology = {
  validationA: {
    events: ['validationEvent'],
    commands: ['validationCommand'],
  },
  validationB: {
    events: ['validationEvent'],
    commands: [],
  },
  validationC: {
    events: [],
    commands: ['validationCommand'],
  },
  validationD: {
    events: [],
    commands: [],
  },
};

const connectionUrls = [
  process.env.AMQP_URL || 'amqp://guest:guest@localhost:5672',
];

const createNewClient = () =>
  createCarrotStiqsClient({
    connectionUrls,
    logger: { log: () => {}, error: () => {}, warn: () => {} },
    topology,
  });

describe('Validations', () => {
  let client = createNewClient();

  // We wait in order to avoid inconsistent errors closing channels.
  beforeEach(() => {
    return client.close().then(() => {
      client = createNewClient();
    });
  });

  afterAll(() => {
    return client.close();
  });

  it("Should throw if topology isn't valid", async () => {
    expect.assertions(8);

    try {
      createCarrotStiqsClient({
        connectionUrls,
        logger: { log: () => {}, error: () => {}, warn: () => {} },
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }

    try {
      createCarrotStiqsClient({
        connectionUrls,
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology: [], // not an object
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }

    try {
      createCarrotStiqsClient({
        connectionUrls,
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology: { fakeGroup: { events: [] } }, // forgot commands
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }

    try {
      createCarrotStiqsClient({
        connectionUrls,
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology: { fakeGroup: { events: [{}], commands: [{}] } }, // arrays should have strings in them
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }
  });

  it("Should throw if connectionUrls isn't valid", async () => {
    expect.assertions(6);

    try {
      createCarrotStiqsClient({
        connectionUrls: [], // must have at least one url
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology,
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }

    try {
      createCarrotStiqsClient({
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology,
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }

    try {
      createCarrotStiqsClient({
        connectionUrls: [{ url: '' }], // not an accepted format
        logger: { log: () => {}, error: () => {}, warn: () => {} },
        topology,
      });
    } catch (err) {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    }
  });

  it("Should reject if group name doesn't exist in topology", async () => {
    expect.assertions(2);

    return client
      .initializeConsumerGroup('notagroup', { events: {}, commands: {} })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if sending to dead letter command is disabled', async () => {
    expect.assertions(1);
    const disableSendingClient = createCarrotStiqsClient({
      connectionUrls,
      logger: { log: () => {}, error: () => {}, warn: () => {} },
      topology: {
        'post-mortem': {
          commands: ['dead-letter'],
          events: [],
        },
      },
    });

    return new Promise((res, rej) => {
      return disableSendingClient
        .initializeConsumerGroup('post-mortem', {
          events: {},
          commands: {
            'dead-letter': {
              prefetch: 1,
              handler: () => {},
            },
          },
        })
        .then(() => {
          client.sendCommand('dead-letter', 'will be rejected').catch((err) => {
            expect(err).toBeInstanceOf(CarrotStiqsError);
            res();
          });
        })
        .catch(rej);
    }).then(() => disableSendingClient.close());
  });

  it('Should reject if any of the registered events for a group are not consumed', () => {
    expect.assertions(2);

    return client
      .initializeConsumerGroup('validationB', {
        events: {},
        commands: {},
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if any of the registered commands for a group are not consumed', () => {
    expect.assertions(2);

    return client
      .initializeConsumerGroup('validationC', {
        events: {},
        commands: {},
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if a group attempts to consume events it has not registered for', () => {
    expect.assertions(2);

    return client
      .initializeConsumerGroup('validationD', {
        events: { notathing: { prefetch: 1, handler: async () => {} } },
        commands: {},
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if a group attempts to consume commands it has not registered for', () => {
    expect.assertions(2);

    return client
      .initializeConsumerGroup('validationD', {
        events: {},
        commands: {
          notathing: {
            prefetch: 1,
            handler: async () => {},
          },
        },
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if attempting to send an event as a command', () => {
    expect.assertions(2);
    return client.sendCommand('validationEvent', 'payload').catch((err) => {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    });
  });

  it('Should reject if attempting to send a command as an event', () => {
    expect.assertions(2);
    return client.sendEvent('validationCommand', 'payload').catch((err) => {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    });
  });

  it('Should reject if attempting to consume an event as a command', () => {
    expect.assertions(2);
    return client
      .initializeConsumerGroup('validationB', {
        events: {
          validationEvent: {
            prefetch: 1,
            handler: async () => {},
          },
        },
        commands: {
          validationEvent: {
            prefetch: 1,
            handler: async () => {},
          },
        },
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if attempting to consume a command as an event', () => {
    expect.assertions(2);
    return client
      .initializeConsumerGroup('validationC', {
        events: {
          validationCommand: {
            prefetch: 1,
            handler: async () => {},
          },
        },
        commands: {
          validationCommand: {
            prefetch: 1,
            handler: async () => {},
          },
        },
      })
      .catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      });
  });

  it('Should reject if attempting to send events that are not registered', () => {
    expect.assertions(2);

    return client.sendEvent('notathing', 'payload').catch((err) => {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    });
  });

  it('Should reject if attempting to send commands that are not registered', () => {
    expect.assertions(2);

    return client.sendCommand('notathing', 'payload').catch((err) => {
      expect(err).toBeInstanceOf(CarrotStiqsError);
      expect(err.message).toMatchSnapshot();
    });
  });

  it('Should throw if event payload is not a string or Buffer', () => {
    expect.assertions(8);

    return Promise.all([
      client.sendEvent('validationEvent', undefined).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendEvent('validationEvent', null).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendEvent('validationEvent', {}).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendEvent('validationEvent', []).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
    ]);
  });

  it('Should throw if command payload is not a string or Buffer', () => {
    expect.assertions(8);

    return Promise.all([
      client.sendCommand('validationCommand', undefined).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendCommand('validationCommand', null).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendCommand('validationCommand', {}).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
      client.sendCommand('validationCommand', []).catch((err) => {
        expect(err).toBeInstanceOf(CarrotStiqsError);
        expect(err.message).toMatchSnapshot();
      }),
    ]);
  });
});
