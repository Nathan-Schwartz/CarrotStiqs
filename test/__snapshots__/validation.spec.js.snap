// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validations Should reject if a group attempts to consume commands it has not registered for 1`] = `"Group validationD attempted to consume commands that were not registered for this group in the topology: notathing"`;

exports[`Validations Should reject if a group attempts to consume events it has not registered for 1`] = `"Group validationD attempted to consume events that were not registered for this group in the topology: notathing"`;

exports[`Validations Should reject if any of the registered commands for a group are not consumed 1`] = `"Group validationC did not provide a handler for commands that were registered for this group in the toplogy: validationCommand"`;

exports[`Validations Should reject if any of the registered events for a group are not consumed 1`] = `"Group validationB did not provide a handler for events that were registered for this group in the toplogy: validationEvent"`;

exports[`Validations Should reject if attempting to consume a command as an event 1`] = `"Group validationC attempted to consume events that were not registered for this group in the topology: validationCommand"`;

exports[`Validations Should reject if attempting to consume an event as a command 1`] = `"Group validationB attempted to consume commands that were not registered for this group in the topology: validationEvent"`;

exports[`Validations Should reject if attempting to send a command as an event 1`] = `"This event does not exist in the topology: validationCommand"`;

exports[`Validations Should reject if attempting to send an event as a command 1`] = `"This command does not exist in the topology: validationEvent"`;

exports[`Validations Should reject if attempting to send commands that are not registered 1`] = `"This command does not exist in the topology: notathing"`;

exports[`Validations Should reject if attempting to send events that are not registered 1`] = `"This event does not exist in the topology: notathing"`;

exports[`Validations Should reject if group name doesn't exist in topology 1`] = `"Attempted to setup consumers for unregistered group: notagroup"`;

exports[`Validations Should throw if command payload is not a string or Buffer 1`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if command payload is not a string or Buffer 2`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if command payload is not a string or Buffer 3`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if command payload is not a string or Buffer 4`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if connectionUrls isn't valid 1`] = `"connectionUrls must be an array of strings."`;

exports[`Validations Should throw if connectionUrls isn't valid 2`] = `"connectionUrls must be an array of strings."`;

exports[`Validations Should throw if connectionUrls isn't valid 3`] = `"connectionUrls must be an array of strings."`;

exports[`Validations Should throw if event payload is not a string or Buffer 1`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if event payload is not a string or Buffer 2`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if event payload is not a string or Buffer 3`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if event payload is not a string or Buffer 4`] = `"Message must be a string or Buffer."`;

exports[`Validations Should throw if topology isn't valid 1`] = `"topology must be an object, where every property has the following type: { events: ['string'], commands: ['string'] }. Either array may be empty."`;

exports[`Validations Should throw if topology isn't valid 2`] = `"topology must be an object, where every property has the following type: { events: ['string'], commands: ['string'] }. Either array may be empty."`;

exports[`Validations Should throw if topology isn't valid 3`] = `"topology must be an object, where every property has the following type: { events: ['string'], commands: ['string'] }. Either array may be empty."`;

exports[`Validations Should throw if topology isn't valid 4`] = `"topology must be an object, where every property has the following type: { events: ['string'], commands: ['string'] }. Either array may be empty."`;
